[PUBDEV-5179] Fix light endpoint - return all columns, but only first…

jakubhava opened the pull request:
… few columns are actually backed by vectors
@mmalohlava This is one possible solution
Advantage -> it's simple
Disadvantage -> when people would really want just a specific number of columns in light endpoint, we anyway return all of them. However, we do not expose this API, for example in PySparkling, which is a single place where this endpoint is used, we do not allow users to set the desired number of columns.

The best solution for the LightEndpoint would still to implement it as LightFrame but that is limited by the Schemas structures (we tried that in the office)

michalkurka commented:
@rpeck please review, you had valid points in our meeting

jakubhava commented:
@michalkurka The current solution works, I think it's important for this to be put into the next release as it's causing real issues in PySparkling. Agree that if there is a better way we should go for it, but I wouldn't like to rush it. 

@rpeck if you please have the ideas how to improve, please let me know :)

rpeck commented:
We should be able to get the functionality you want with the existing REST API, with one improvement and perhaps a bug fix that we should implement anyway.

- {{/3/Frames}} has pagination built in, so you can specify a rectangle of rows/columns to return; Flow uses this for pagination of frame preview
- the entire REST API has an {{exclude_fields}} parameter available.  By default this works by generating the whole payload and then filtering out the unneeded fields, but the individual handlers can be smarter and not generate the unneeded fields, if they are expensive to generate as in this case
- it's possible that there's a bug / possible important improvement in the {{FramesHandler}}: it might trigger frame summaries of all columns even if only some are requested via pagination.  If this bug is real then it should be fixed to speed up {{Frame}} preview in {{Flow}}.

So to implement what you need by improving the {{/3/Frames}} REST API you'd have the client call {{/3/Frames}} twice: once excluding the data field entirely, to get all the names, and once with column pagination to get the preview for the columns you need.  Make the handler a bit smarter so that it doesn't compute rollups / summaries for columns that aren't needed, and doesn't compute anything for the data preview if it's not needed.

jakubhava commented:
I just read this again properly, sorry for delays. The proposal makes sense, however, it doesn't seem to be applicable to this case.

We introduced the light frame API to be able to get the preview of data, but only with a specified number of columns and rows. This is still fine and current API can be used for that. What we need to achieve by this change is to obtain all column names, but load data preview for just a specified amount of first n columns.

The code wasn't that much obvious so I improved it a little bit. Hopefully, it makes more sense now. I introduced a new field called `full_column_count` which basically specifies how many columns need to be returned with backing data. This can be used in cases where the user has a lot of columns, wants to have the frame preview referencing all columns (for example, the user wants to run algo on all columns) but for performance optimizations, the user just asks for preview of `full_column_count` columns.

@michalkurka @rpeck what do you think? I would rather fix the possible bug @rpeck reported in a different jira.

jakubhava commented:
I just read this again properly, sorry for delays. The proposal makes sense, however, it doesn't seem to be applicable to this case.

We introduced the light frame API to be able to get the preview of data, but only with a specified number of columns and rows. This is still fine and current API can be used for that. What we need to achieve by this change is to obtain all column names, but load data preview for just a specified amount of first n columns.

The code wasn't that much obvious so I improved it a little bit. Hopefully, it makes more sense now. I introduced a new field called `full_column_count` which basically specifies how many columns need to be returned with backing data. This can be used in cases where the user has a lot of columns, wants to have the frame preview referencing all columns (for example, the user wants to run algo on all columns) but for performance optimizations, the user just asks for preview of `full_column_count` columns.

@michalkurka @rpeck what do you think? I would rather fix the possible bug @rpeck reported in a different jira.

michalkurka commented on h2o-core/src/main/java/water/api/schemas3/FrameV3.java:
```diff
@@ -261,11 +284,18 @@ private FrameV3 fillFromImpl(Frame f, long row_offset, int row_count,
-      else
-        columns[i] = new ColV3(f._names[column_offset + i], vecs[column_offset + i], this.row_offset, this.row_count);
+      else {
+        if(i < full_column_count) {
```
This can be simplified - the condition can be in the last argument, 'if' is not necessary

jakubhava replied:
Good point, thanks!

michalkurka commented:
@jakubhava I think the change looks fine. If we introduce the new API (full_column), will this still be relevant when we do @rpeck's proposed refactoring? Or is it only needed for the fix?

I think I agree we should primarily fix the bug and leave the refactoring for later but changing the API back and forth is also not ideal.

jakubhava commented:
@michalkurka The @rpeck proposal would not solve this particular issue. We can select frame slice, but want we need to do in this case is to specify how many columns should be obtained together with data. This feature is missing in the code,  so I think it can be a good addition anyway. 

> On 5 Feb 2018, at 17:53, michalkurka <notifications@github.com> wrote:
> 
> @jakubhava I think the change looks fine. If we introduce the new API (full_column), will this still be relevant when we do @rpeck's proposed refactoring? Or is it only needed for the fix?
> 
> I think I agree we should primarily fix the bug and leave the refactoring for later but changing the API back and forth is also not ideal.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jakubhava commented:
Few tests were failing, should be fixed now. I forgot to set the default value for the new parameter. Right now it does not affect anything if the user does not explicitly specifies it.

From my point of view, this change is good to go. Let me please know if you have any other proposals

jakubhava commented:
@michalkurka @rpeck I would like to merge this one so it can be in the next release -> and next release of SW. Would that be ok for you?

jakubhava commented:
Thank you for the review!
